beetsplug/fish.py

```diff
@@ -0,0 +1,245 @@
+    for cmd in cmds:
+        name = cmd.name
+        word += ("\n" * 2) + "# ====== {} =====".format(
+            "completions for  " + cmd.name) + "\n"
+        if cmd.aliases:
+            word += "# === {} =".format(
+                "aliases for " + cmd.name) + "\n"
+            for alias in cmd.aliases:
+                alias = alias if alias != "?" else "\?"
+                word += "abbr -a {} {}".format(
+                    alias, name) + "\n"
+
+        word += "\n"
+
+        for option in cmd.parser._get_all_options()[1:]:
+
+            cmd_LO = (" -l " + option._long_opts[0].replace('--', '')
+                      )if option._long_opts else ''
+            cmd_SO = (" -s " + option._short_opts[0].replace('-', '')
+                      ) if option._short_opts else ''
+            cmd_needARG = ' -r ' if option.nargs in [1] else ''
+            cmd_helpstr = (" -d " + wrap(' '.join(option.help.split()))
+                           ) if option.help else ''
+            cmd_arglist = (' -a ' + wrap(" ".join(option.choices))
+                           ) if option.choices else ''
+
+            word += " ".join(bl_use3.format(
+                name,
+                (cmd_needARG + cmd_SO + cmd_LO + " -f " + cmd_arglist),
+                cmd_helpstr).split()) + "\n"
+
+        word = (word + " ".join(bl_use3.format(
+            name,
+            ("-s " + "h " + "-l " + "help" + " -f "),
+            ('-d ' + wrap("print help") + "\n")
+        ).split()))
+    return word
+
+
+def get_fishfile(opts, lib):
+    cmds = sorted((commands.default_commands +
+                  commands.plugins.commands()),  key=attrgetter('name'))
+    cmds_helps = [cmd.help for cmd in cmds]
+    fields = sorted(set(library.Album.all_keys() + library.Item.all_keys()))
+
+    totstring = head
+
+    cmds_names = [cmd.name for cmd in cmds]
+    totstring += (
+        "set CMDS " + " ".join(cmds_names) + ("\n" * 2)
+    )
+    fields = (field + ":" for field in fields)
+    if opts.noFields is False:
+        totstring += (
+            "set FIELDS " + " ".join(fields) + ("\n" * 2)
+        )
+    if opts.extraFields:
+        setOfValues = getSetOfValuesForField(lib,  opts.extraFields)
+        for fld in opts.extraFields:
+            extraname = fld.upper() + 'S'
+            totstring += (
+                "set  " + extraname + " " + " ".join(sorted(setOfValues[fld]))
+                + ("\n" * 2)
+            )
+    # Collect subcommands
+    totstring += ("\n" * 2) + "# ====== {} =====".format(
+        "basic beet options") + ("\n" * 2)
+
+    totstring += getBasicBeetOptions()
+
+    totstring += ("\n" * 2) + "# ====== {} =====".format(
+        "basic beet subarguments") + ("\n" * 2)
+
+    cmds_zipped = zip(cmds_names, cmds_helps)
+    totstring += getSubarguments(cmds_zipped, opts)
+
+    totstring += "\n"
+    totstring += getAllCommands(cmds)
+
+    return totstring
+
+bl_need2 = """complete -c beet -n '__fish_beet_needs_command' {} {}\n"""
```

Comment:
Constants like these, by convention, go at the top of the file and get CAPITALIZED names, like `BL_NEED2`.